wrappers/internal/service_unit_gen.go

```diff
@@ -292,25 +297,31 @@ WantedBy={{.ServicesTarget}}
 		OOMAdjustScore: oomAdjustScore,
 		BusName:        busName,
 
-		Before: generateServiceNames(appInfo.Snap, appInfo.Before),
-		After:  generateServiceNames(appInfo.Snap, appInfo.After),
+		Before:  generateServiceNames(appInfo.Snap, appInfo.Before),
+		After:   generateServiceNames(appInfo.Snap, appInfo.After),
+		BindsTo: generateServiceNames(appInfo.Snap, appInfo.BindsTo),
 
 		// systemd runs as PID 1 so %h will not work.
 		Home: "/root",
 	}
-	switch appInfo.DaemonScope {
-	case snap.SystemDaemon:
+	switch {
+	case appInfo.DaemonScope.IsSystemDaemon():
 		wrapperData.ServicesTarget = systemd.ServicesTarget
 		wrapperData.PrerequisiteTarget = systemd.PrerequisiteTarget
 		wrapperData.MountUnit = filepath.Base(systemd.MountUnitPath(dirs.StripRootDir(appInfo.Snap.MountDir())))
 		wrapperData.Requires = append(wrapperData.Requires, wrapperData.MountUnit)
 		wrapperData.WorkingDir = dirs.StripRootDir(appInfo.Snap.DataDir())
 		wrapperData.After = append(wrapperData.After, "snapd.apparmor.service")
-	case snap.UserDaemon:
+	case appInfo.DaemonScope.IsUserDaemon():
 		wrapperData.ServicesTarget = systemd.UserServicesTarget
 		// FIXME: ideally use UserDataDir("%h"), but then the
 		// unit fails if the directory doesn't exist.
 		wrapperData.WorkingDir = appInfo.Snap.DataDir()
+		if appInfo.DaemonScope == snap.GraphicalUserDaemonScope {
+			wrapperData.After = append(wrapperData.After, "graphical-session.target")
+			wrapperData.BindsTo = append(wrapperData.BindsTo, "graphical-session.target")
```

Comment:
There's the risk here that we're generalizing the kind of service too much imho, in fact depending on the role of the user service, we could use either `BindsTo` or `PartOf`.

Where a service may need to be put down with graphical session being closed, but it not necessarily needs to be started at startup (e.g. it's dbus-activated).

See the comment on [`systend.special`](https://manpages.ubuntu.com/manpages/questing/en/man7/systemd.special.7.html) in `Special Passive User Units`.

Also here it would be probably important (mostly for core-desktop) to consider that there's even the `graphical-session-pre.target` case

Reply:
So... if the unit is dbus-activated, we should use `PartOf` instead of `BindsTo`?

The pre-session can easily be implemented with an extra scope, but at this moment I don't know any specific case that requires it.

Reply:
> So... if the unit is dbus or socket activated, we should use PartOf instead of BindsTo?

Yes, as you don't want to start it by default, only when triggered by a dbus request.

> so I think that we should concentrate now into merging this

Sure, but I think it's better to try to design things in away that can be improved in the future.